src/vos/vos_common.c

```diff
@@ -776,57 +780,70 @@ vos_self_init(const char *db_path, bool use_sys_db, int tgt_id)
 {
 	char		*evt_mode;
 	int		 rc = 0;
-	char		*lmm_db_path = getenv("DAOS_LMM_DB_PATH");
+	char		*lmm_db_path = NULL;
 	struct sys_db	*db;
 
 	D_MUTEX_LOCK(&self_mode.self_lock);
 	if (self_mode.self_ref) {
 		self_mode.self_ref++;
-		D_GOTO(out, rc);
+		goto out;
 	}
 
 	rc = ABT_init(0, NULL);
 	if (rc != 0)
-		D_GOTO(out, rc);
+		goto out;
 
 	vos_start_epoch = 0;
 
 #if VOS_STANDALONE
 	self_mode.self_tls = vos_tls_init(0, -1);
 	if (!self_mode.self_tls) {
 		ABT_finalize();
-		D_GOTO(out, rc);
+		goto out;
 	}
 #endif
 	rc = vos_mod_init();
 	if (rc)
-		D_GOTO(failed, rc);
+		goto failed;
+
+	if (bio_nvme_configured(SMD_DEV_TYPE_META)) {
+		bool	needs_free = false;
+
+		rc = bio_get_lmm_db_path(lmm_db_path);
+		if (rc) {
+			if (rc == -DER_INVAL)
+				lmm_db_path = LMMDB_PATH;
+			else
+				goto failed;
+		} else {
+			needs_free = true;
+		}
 
-	if (!bio_nvme_configured(SMD_DEV_TYPE_META)) {
-		if (use_sys_db)
-			rc = vos_db_init(db_path);
-		else
-			rc = vos_db_init_ex(db_path, "self_db", true, true);
-		db = vos_db_get();
-	} else {
-		if (lmm_db_path == NULL)
-			lmm_db_path = LMMDB_PATH;
 		if (use_sys_db)
 			rc = lmm_db_init(lmm_db_path);
 		else
 			rc = lmm_db_init_ex(lmm_db_path, "self_db", true, true);
 		db = lmm_db_get();
+
+		if (needs_free)
+			D_FREE(lmm_db_path);
+	} else {
+		if (use_sys_db)
+			rc = vos_db_init(db_path);
+		else
+			rc = vos_db_init_ex(db_path, "self_db", true, true);
+		db = vos_db_get();
 	}
 	if (rc)
-		D_GOTO(failed, rc);
+		goto failed;
 
 	rc = smd_init(db);
 	if (rc)
-		D_GOTO(failed, rc);
+		goto failed;
 
 	rc = vos_self_nvme_init(db_path, tgt_id);
```

Comment:
Since you already touched this code, it is good idea to move nvme init before sys db init, otherwise, we don't know if MD-ON-SSD configured or not.

Reply:
yes good point, that needs to be called before bio_nvme_configured(META) because bio_spdk_env_init needs to be called to populate roles, thanks

Reply:
done